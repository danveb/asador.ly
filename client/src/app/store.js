import { configureStore } from "@reduxjs/toolkit";
import authReducer from "../redux/auth/authSlice";
import pinReducer from "../redux/pins/pinSlice";

export const store = configureStore({
  reducer: {
    auth: authReducer,
    pins: pinReducer,
  },
}); 