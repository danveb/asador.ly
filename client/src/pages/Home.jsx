import { Footer } from "../components";
import { Link } from "react-router-dom";
import "../styles/Home.scss";

const Home = () => {
  return (
    <>
      <div className="home">
        <div className="home__main">
          <div className="home__title">
            <h2>Welcome to Asador</h2>
            <p>Share the best parrillas (steakhouses) of Buenos Aires. From traditional eateries to world-acclaimed establishments, there's always a fix for authentic Argentine <span>asado</span>.</p>
            <Link to="/map">Check Our Map</Link>
          </div>
        </div>
        <div className="home__content">
          <div className="home__subtitle">
            <h2>Asador</h2>
            <p>At <span>Asador</span> we have curated a list of the best parrillas in Buenos Aires -- we may have missed a few, on purpose, just so you can share your favorite ones. </p>
          </div>
        </div>
      </div>
      <Footer />
    </>
  )
}

export default Home