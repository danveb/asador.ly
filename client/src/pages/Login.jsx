import { useState, useEffect } from "react";
import { Link, useNavigate } from "react-router-dom";
import { Spinner } from "../components/index";
import { useSelector, useDispatch } from "react-redux";
import { login, reset } from "../redux/auth/authSlice";
import { toast } from "react-toastify";
import "../styles/Login.scss";

const Login = () => {
  const [formData, setFormData] = useState({
    username: "",
    password: "",
  });

  // destructure formData 
  const { username, password } = formData;

  // useNavigate 
  const navigate = useNavigate();

  // useDispatch
  const dispatch = useDispatch();

  const { user, isLoading, isError, isSuccess, message } = useSelector((state) => state.auth);

  // useEffect 
  useEffect(() => {
    if (isError) {
      toast.error(message, {
        position: "top-center",
        autoClose: 2000,
        pauseOnHover: true,
      });
    };
    if (isSuccess) {
      toast.success(`Welcome back ${user.username} ðŸ‘‹ `, {
        position: "top-center",
        autoClose: 1500,
        hideProgressBar: true,
      });
      navigate("/map");
    };
    // redirect existing user back to map
    if (user) {
      navigate("/map");
    };
    dispatch(reset());
    // dependency   
  }, [user, isError, isSuccess, message, navigate, dispatch]);

  // handleChange
  const handleChange = (e) => {
    setFormData((prevState) => ({
      ...prevState,
      [e.target.name]: e.target.value
    }));
  };

  // handleSubmit 
  const handleSubmit = (e) => {
    e.preventDefault();
    const userData = {
      username,
      password,
    };
    // // error checking with toast... 
    if (username.length < 3) {
      toast.error("Username incorrect. Minimum length is 3 characters.");
    } else if (password.length < 4) {
      toast.error("Password incorrect. Minimum length is 4 characters.");
    } else {
      dispatch(login(userData));
    }
  };

  if (isLoading) {
    return <Spinner />
  };

  return (
    <div className="login">
      <div className="login__wrapper">
        <div className="login__head">
          <h1>Login</h1>
          <p>Login and start pinning your favorite <span>parrillas</span> in Buenos Aires.</p>
        </div>
        <form className="form" onSubmit={handleSubmit}>
          <label
            htmlFor="username"
          >
            Username</label>
          <input
            id="username"
            name="username"
            value={username}
            onChange={handleChange}
            type="text"
            placeholder="Your username"
            autoComplete="username"
            required
          />
          <label
            htmlFor="password"
          >
            Password</label>
          <input
            id="password"
            name="password"
            value={password}
            onChange={handleChange}
            type="password"
            placeholder="Your password"
            autoComplete="off"
            required
          />
          <button>Login</button>
          <div className="register__link">
            <Link to="/register">Don't have an account?</Link>
          </div>
        </form>
      </div>
    </div>
  )
}

export default Login