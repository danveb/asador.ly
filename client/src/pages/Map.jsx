import { useEffect, useState } from "react";
import Spinner from "../components/Spinner";
import { Map as Mapbox, Marker, Popup } from "react-map-gl";
import RoomIcon from '@mui/icons-material/Room';
import StarRateIcon from '@mui/icons-material/StarRate';
import AttachMoneyIcon from '@mui/icons-material/AttachMoney';
import { useSelector, useDispatch } from "react-redux";
import { createPin, getPins, reset } from "../redux/pins/pinSlice";
import "../styles/Map.scss";
import { toast } from "react-toastify";

/* 
 * NOTE: 
 * Using react-map-gl to render mapbox map (v7.0.18)
 * Still need to install mapbox-gl (v2.9.2) to use mapbox stylesheet
*/

const Map = () => {
  const { user } = useSelector((state) => state.auth);
  const { pins, isLoading, isError, message } = useSelector((state) => state.pins);

  // useDispatch
  const dispatch = useDispatch();

  // useState
  const [viewState, setViewState] = useState({
    // main coordinates of Buenos Aires
    longitude: -58.403578,
    latitude: -34.608056,
    zoom: 11.5
  });
  const [showPopup] = useState(true);
  const [currentPlaceId, setCurrentPlaceId] = useState(null);
  const [newPlace, setNewPlace] = useState(null);
  const [formData, setFormData] = useState({
    title: "",
    description: "",
    rating: "",
    cost: "",
  });

  const { title, description, rating, cost } = formData;

  const handleChange = (e) => {
    setFormData((prevState) => ({
      ...prevState,
      [e.target.name]: e.target.value
    }));
  };

  // useEffect
  useEffect(() => {
    if (isError) {
      toast.error("Sorry, there was an error. Retrying", {
        position: "top-center",
        autoClose: 2500,
        pauseOnHover: false,
      });
    }
    dispatch(getPins());
  }, [dispatch, isError, message]);

  // handleMarkerClick
  const handleMarkerClick = (id, long, lat) => {
    setCurrentPlaceId(id);
    setViewState({
      ...viewState,
      longitude: long,
      latitude: lat,
    });
  };

  // handleAddClick 
  const handleAddClick = (e) => {
    if (user) {
      const latitude = e.lngLat.lat;
      const longitude = e.lngLat.lng;
      setNewPlace({
        lat: latitude,
        lng: longitude,
      });
    } else {
      toast("Would you like to create an account?", {
        position: "top-center",
        autoClose: 3500,
        pauseOnHover: false,
      });
    };
  };

  // handleSubmit (REDUX) 
  const handleSubmit = (e) => {
    e.preventDefault();
    const pinData = {
      username: user.username,
      title,
      description,
      rating,
      cost,
      location: {
        coordinates: [newPlace.lng, newPlace.lat]
      },
    };
    dispatch(createPin(pinData));
    setFormData({
      title: "",
      description: "",
      rating: "",
      cost: "",
    });
    setNewPlace(null);
    dispatch(reset());
    toast.success("Pin added ðŸ¥³ ", {
      position: "top-center",
      autoClose: 3000,
      pauseOnHover: false,
    });
  };

  if (isLoading) {
    return <Spinner />
  };

  return (
    <div className="mainMap">
      <Mapbox
        // controlled map (state)
        {...viewState}
        onMove={evt => setViewState(evt.viewState)}
        style={{ width: "100%", height: "100%" }}
        // mapStyle="mapbox://styles/mapbox/streets-v9"
        // custom map style
        mapStyle="mapbox://styles/danvebb/cl6frln64000414mrhya0xc3k"
        // mapbox token
        mapboxAccessToken={process.env.REACT_APP_MAPBOX}
        // double click to create a pin 
        onDblClick={handleAddClick}
        transition="300"
      >
        {pins.map((pin) => (
          <div key={pin._id}>
            <Marker
              longitude={pin.location.coordinates[0]}
              latitude={pin.location.coordinates[1]}
              anchor="bottom"
              offsetLeft={-viewState.zoom * 3.5}
              offsetTop={-viewState.zoom * 7}
            >
              <RoomIcon
                style={{ color: "crimson", fontSize: viewState.zoom * 4, cursor: "pointer" }}
                onClick={() => handleMarkerClick(pin._id, pin.location.coordinates[0], pin.location.coordinates[1])}
              />
            </Marker>
            {showPopup && pin._id === currentPlaceId && (
              <Popup
                longitude={pin.location.coordinates[0]}
                latitude={pin.location.coordinates[1]}
                anchor="top"
                closeOnClick={false}
                onClose={() => setCurrentPlaceId(null)}
              >
                <div className="card">
                  <label>Place</label>
                  <h4 className="place">{pin.title}</h4>
                  <label>Description</label>
                  <p className="description">{pin.description}</p>
                  <label>Rating</label>
                  <div className="icons">
                    {Array(pin.rating).fill(<StarRateIcon className="star" />)}
                  </div>
                  <label>Cost</label>
                  <div className="icons">
                    {Array(pin.cost).fill(<AttachMoneyIcon className="dollar" />)}
                  </div>
                  <label>Info</label>
                  <span className="username">Created by <b>{pin.username}</b></span>
                </div>
              </Popup>
            )}
          </div>
        ))}
        {showPopup && newPlace && (
          <Popup
            longitude={newPlace.lng}
            latitude={newPlace.lat}
            anchor="top"
            closeOnClick={false}
            onClose={() => setNewPlace(null)}
          >
            <div>
              <form onSubmit={handleSubmit}>
                <label
                  htmlFor="title"
                >Title</label>
                <input
                  id="title"
                  name="title"
                  value={title}
                  placeholder="Enter Title"
                  onChange={handleChange}
                  required
                />
                <label
                  htmlFor="description"
                >Description</label>
                <textarea
                  id="description"
                  name="description"
                  value={description}
                  placeholder="Want To Share About This Steakhouse?"
                  onChange={handleChange}
                />
                <label
                  htmlFor="rating"
                >Rating</label>
                <select
                  id="rating"
                  name="rating"
                  value={rating}
                  onChange={handleChange}
                >
                  <option defaultValue={"default"}>Choose an option</option>
                  <option value={"1"}>1</option>
                  <option value={"2"}>2</option>
                  <option value={"3"}>3</option>
                  <option value={"4"}>4</option>
                  <option value={"5"}>5</option>
                </select>
                <label
                  htmlFor="cost"
                >Cost</label>
                <select
                  id="cost"
                  name="cost"
                  value={cost}
                  onChange={handleChange}
                >
                  <option defaultValue={"default"}>Choose an option</option>
                  <option value={"1"}>1</option>
                  <option value={"2"}>2</option>
                  <option value={"3"}>3</option>
                  <option value={"4"}>4</option>
                  <option value={"5"}>5</option>
                </select>
                <button className="submit-btn" type="submit">Add Pin</button>
              </form>
            </div>
          </Popup>
        )}
      </Mapbox>
    </div>
  )
}

export default Map