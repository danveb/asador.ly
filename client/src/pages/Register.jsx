import { useState, useEffect } from "react";
import { Link, useNavigate } from "react-router-dom";
import { Spinner } from "../components/index";
import { useSelector, useDispatch } from "react-redux";
import { register, reset } from "../redux/auth/authSlice";
import { toast } from "react-toastify";
import "../styles/Register.scss";

const Register = () => {
  const [formData, setFormData] = useState({
    username: "",
    email: "",
    password: "",
  });

  // destructure formData 
  const { username, email, password } = formData;

  // useNavigate 
  const navigate = useNavigate();

  // useDispatch
  const dispatch = useDispatch();

  const { user, isLoading, isError, isSuccess, message } = useSelector((state) => state.auth);

  // useEffect 
  useEffect(() => {
    if (isError) {
      toast.error(message, {
        position: "top-center",
        autoClose: 2000,
        pauseOnHover: true,
      });
    };
    if (isSuccess) {
      toast.success(`Welcome ${user.username} ðŸŽ‰ `, {
        position: "top-center",
        autoClose: 1500,
        pauseOnHover: false,
      });
      navigate("/map");
    };
    // redirect existing user back home
    if (user) {
      navigate("/");
    };
    dispatch(reset());
    // dependency   
  }, [user, isError, isSuccess, message, navigate, dispatch]);

  // handleChange
  const handleChange = (e) => {
    setFormData((prevState) => ({
      ...prevState,
      [e.target.name]: e.target.value
    }));
  };

  // handleSubmit 
  const handleSubmit = (e) => {
    e.preventDefault();
    const userData = {
      username,
      email,
      password
    };
    // error checking with toast... 
    if (username.length < 3) {
      toast.error("Please create a new username. Minimum length is 3 characters.")
    } else if (password.length < 4) {
      toast.error("Please create a new password. Minimum length is 4 characters.");
    } else {
      dispatch(register(userData));
    };
  };

  if (isLoading) {
    return <Spinner />
  };

  return (
    <div className="register">
      <div className="register-top">
        <div className="register-title">
          <h1>Register</h1>
          <p>Create an account to share with us your favorite <span>parrillas</span> in Buenos Aires.</p>
        </div>
        <form className="register-form" onSubmit={handleSubmit}>
          <label
            htmlFor="username"
          >Username</label>
          <input
            id="username"
            name="username"
            value={username}
            onChange={handleChange}
            type="text"
            placeholder="Create a username"
            required
          />
          <label
            htmlFor="email"
          >Email</label>
          <input
            id="email"
            name="email"
            value={email}
            onChange={handleChange}
            type="email"
            placeholder="Your email address"
            required
          />
          <label
            htmlFor="password"
          >Password</label>
          <input
            id="password"
            name="password"
            value={password}
            onChange={handleChange}
            type="password"
            placeholder="Your password"
            required
          />
          <button>Register</button>
          <Link to="/login">Already have an account?</Link>
        </form>
      </div>
    </div>
  )
}

export default Register