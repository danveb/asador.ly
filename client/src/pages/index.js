export { default as Login } from "./Login";
export { default as Home } from "./Home";
export { default as Map } from "./Map";
export { default as Register } from "./Register"; 