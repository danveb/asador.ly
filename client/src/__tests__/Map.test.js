import { render } from "@testing-library/react";
import { store } from "../app/store";
import { Provider } from "react-redux";
import { Map } from "../pages";

describe("Map component", () => {
  // smoke test
  test("renders Map component", () => {
    render(
      <Provider store={store}>
        <Map />
      </Provider>
    );
  });

  // snapshot test
  test("matches snapshot", () => {
    const { asFragment } = render(
      <Provider store={store}>
        <Map />
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});