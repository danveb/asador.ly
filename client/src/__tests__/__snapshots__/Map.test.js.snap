// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Map component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="spinner"
  >
    <div
      class="spinner__loading"
    />
  </div>
</DocumentFragment>
`;
